int-set/src/lib.rs

```diff
@@ -12,45 +12,92 @@
 //! way to create the set is to initialize it from a sorted (ascending) iterator of the values.
+//!
+//! For a type to be stored in the IntSet it must implement the IntSetMember trait, and all
+//! unique values of that type must be able to be mapped to and from a unique u32 value.
+//! See the IntSetMember trait for more information.
 
 mod bitpage;
 mod bitset;
 
 use bitset::BitSet;
 use std::hash::Hash;
+use std::marker::PhantomData;
 use std::ops::RangeInclusive;
 
 /// A fast & efficient invertible ordered set for small (up to 32-bit) unsigned integer types.
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
-pub struct IntSet<T>(Membership<T>);
+pub struct IntSet<T>(Membership, PhantomData<T>);
+
+/// Members of IntSet must implement this trait. Members of IntSet's must meet the following
+/// conditions to be used in an IntSet:
+///
+/// 1. Every possible unique value of T must be able map to and from a unique u32
+///    integer.
+///
+/// 2. The mapped u32 values must retain the same ordering as the values in T.
+///
+/// 3. all_values_iter() must iterate over all values in T in sorted order (ascending).
+///
+/// from_mapped() will only ever be called with u32 values that were generated by a
+/// call to to_mapped() with a valid value of T. So it doesn't need to correctly handle values
+/// that are outside the domain of T.
+pub trait IntSetMember<T> {
+    fn to_mapped(&self) -> u32;
+    fn from_mapped(member: InDomain) -> T;
+    fn all_values_sorted_iter() -> impl Iterator<Item = u32>;
+}
+
+/// Marks a mapped value as being in the domain of T for IntSetMember<T>. See IntSetMember
+/// for more information.
+pub struct InDomain {
+    value: u32,
+}
 
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
-enum Membership<T> {
+enum Membership {
     /// Records a set of integers which are members of the set.
-    Inclusive(BitSet<T>),
+    Inclusive(BitSet),
 
     /// Records the set of integers which are not members of the set.
-    Exclusive(BitSet<T>),
+    Exclusive(BitSet),
 }
 
-impl<T: Into<u32> + Copy> Default for IntSet<T> {
+impl InDomain {
+    pub fn value(&self) -> u32 {
+        self.value
+    }
+}
+
+impl<T: IntSetMember<T>> Default for IntSet<T> {
     fn default() -> IntSet<T> {
         IntSet::empty()
     }
 }
 
-impl<T: Into<u32> + Copy> IntSet<T> {
+impl<T: IntSetMember<T>> IntSet<T> {
     // TODO(garretrieger): add additional functionality that the harfbuzz version has:
     // - Iteration in reverse.
     // - Iteration starting from some value (and before some value for reverse).
     // - Set operations (union, subtract, intersect, sym diff).
     // - Intersects range and intersects iter.
     // - min()/max()
 
+    /// Returns an iterator over all members of the set.
+    /// Note: iteration of inverted sets can be extremely slow due to the very large number of members in the set
+    /// care should be taken when using .iter() in combination with an inverted set.
```

Comment:
Do you mean it's slower per item or merely that because the inversion probably has a lot of entries it'll be slow?

Reply:
The latter, if the domain of T is all u32's iterating through them all takes a pretty long time even if the individual next() operations are fast. In harfbuzz we found if you're trying to iterate through an inverted set there's usually a better way to set up the algorithm that avoids the iteration in the inverted case.